components/brave_extension/extension/brave_extension/background/today/feedToData.ts

```diff
@@ -0,0 +1,260 @@
+  if (!headlines.length) {
+    return null
+  }
+
+  const articlesWithCategory = featuredCategory
+    ? take(
+        articles,
+        3,
+        a => a.category === featuredCategory,
+      )
+    : []
+
+  let deals = !dealsCategory
+    ? []
+    : take(allDeals, 3, d => d.offers_category === dealsCategory)
+  if (deals.length < 3) {
+    deals.concat(allDeals.splice(0, 3 - deals.length))
+  }
+
+  const publisherInfo = generateArticleSourceGroup(articles)
+
+  const randomArticles = take(articles, 4, isArticleWithin48Hours, true)
+
+  return {
+    articles: headlines,
+    randomArticles,
+    deals,
+    itemsByCategory: (featuredCategory && articlesWithCategory.length) ? { categoryName: featuredCategory, items: articlesWithCategory } : undefined,
+    itemsByPublisher: publisherInfo ? { name: publisherInfo[0], items: publisherInfo[1] } : undefined,
+  }
+}
+
+function generateArticleSourceGroup (articles: BraveToday.Article[]): [string, BraveToday.Article[]] | undefined {
+  const firstArticleWithSource = articles.find(a => !!a.publisher_id)
+  if (firstArticleWithSource) {
+    return [
+      firstArticleWithSource.publisher_id,
+      take(
+        articles,
+        3,
+        a => a.publisher_id === firstArticleWithSource.publisher_id,
+      )
+    ]
+  }
+  return undefined
+}
+
+function take<T>(items: T[], count: number, matching?: (item: T) => boolean, random: boolean = false): T[] {
+  let indicesToTake: number[] = []
+  for (const [i, item] of items.entries())  {
+    const shouldTake = matching ? matching(item) : true
+    if (!shouldTake) {
+      continue
+    }
+    // Take item
+    indicesToTake.push(i)
+    // Stop, if we're limiting the count
+    if (!random && count && indicesToTake.length === count) {
+      break
+    }
+  }
+  // Randomize
+  if (random) {
+    shuffleArray(indicesToTake)
+    if (count && indicesToTake.length > count) {
+      indicesToTake = indicesToTake.slice(0, count)
+    }
+  }
+  // Get items and remove from array
+  indicesToTake = indicesToTake.reverse()
+  const takenItems: T[] = []
+  for (const i of indicesToTake) {
+    takenItems.push(...items.splice(i, 1))
+  }
+  return takenItems.reverse()
+}
+
+function domainFromUrlString(urlString: string): string {
+  return new window.URL(urlString).host
+}
+
+async function weightArticles(articles: BraveToday.FeedItem[]): Promise<BraveToday.FeedItem[]> {
+  // hosts from latest max-200 history items
+  const historyItems: chrome.history.HistoryItem[] = chrome.history
+    ? await new Promise(
+        resolve => chrome.history.search({ text: '', maxResults: 200 }, resolve)
+      )
+    : []
+  const historyHosts: string[] = historyItems.map(h => h.url).filter(i => i).map((u: string) => domainFromUrlString(u))
+  for (const article of articles) {
+    let score = article.score
+    const hasDomainHistoryMatch = historyHosts.includes(domainFromUrlString(article.url))
+    if (hasDomainHistoryMatch) {
+      score -= 5
```

Comment:
What is the significance of 5? (what is a max score? etc)

Reply:
Spec determines 5. iOS has the same hardcoded value. Works with the initial score determined server-side.